File: app/src/main/java/online/fatbook/fatbookapp/ui/activity/SplashActivity.java
package online.fatbook.fatbookapp.ui.activity;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.os.Handler;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;

import online.fatbook.fatbookapp.R;
import online.fatbook.fatbookapp.core.User;
import online.fatbook.fatbookapp.databinding.ActivitySplashBinding;
import online.fatbook.fatbookapp.retrofit.RetrofitFactory;
import online.fatbook.fatbookapp.ui.viewmodel.UserViewModel;
import online.fatbook.fatbookapp.util.UserUtils;

import org.apache.commons.lang3.StringUtils;

import java.util.logging.Level;

import lombok.extern.java.Log;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

@Log
public class SplashActivity extends AppCompatActivity {

    private ActivitySplashBinding binding;

    private UserViewModel userViewModel;

    private String userLogin;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivitySplashBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        userViewModel = new ViewModelProvider(this).get(UserViewModel.class);

        SharedPreferences sharedPreferences = getSharedPreferences(UserUtils.APP_PREFS, Context.MODE_PRIVATE);
        userLogin = sharedPreferences.getString(UserUtils.USER_LOGIN, StringUtils.EMPTY);
        if (StringUtils.isEmpty(userLogin)) {
            userViewModel.setUser(new User());
        } else {
            loadUserData(userLogin);
        }

        userViewModel.getUser().observe(this, user -> {
            Handler handler = new Handler();
            handler.postDelayed(() -> {
                Intent intent;
                if (StringUtils.isEmpty(userLogin)) {
                    intent = new Intent(this, WelcomeActivity.class);
                } else {
                    intent = new Intent(this, MainActivity.class);
                }
                intent.putExtra(UserUtils.TAG_USER, user);
                startActivity(intent);
                finish();
            }, 1);
        });

        binding.buttonSplashRetry.setOnClickListener(view -> {
            binding.textViewSplashError.setVisibility(View.GONE);
            binding.buttonSplashRetry.setVisibility(View.GONE);
            loadUserData(userLogin);
        });
    }

    private void loadUserData(String login) {
        RetrofitFactory.apiServiceClient().getUser(login).enqueue(new Callback<User>() {
            @Override
            public void onResponse(@NonNull Call<User> call, @NonNull Response<User> response) {
                log.log(Level.INFO, "" + response.code() + " found user: " + response.body());
                userViewModel.setUser(response.body());
            }

            @Override
            public void onFailure(@NonNull Call<User> call, @NonNull Throwable t) {
                log.log(Level.INFO, "load user ERROR");
                if (t.toString().contains("ConnectException")) {
                    binding.textViewSplashError.setText(getString(R.string.splash_no_connection_error_client));
                } else {
                    binding.textViewSplashError.setText(getString(R.string.splash_no_connection_error_api));
                }
                binding.textViewSplashError.setVisibility(View.VISIBLE);
                binding.buttonSplashRetry.setVisibility(View.VISIBLE);
            }
        });
    }
}

File: app/src/main/java/online/fatbook/fatbookapp/ui/activity/SplashActivity.kt
package online.fatbook.fatbookapp.ui.activity

import android.content.Intent
import android.content.SharedPreferences
import android.os.Bundle
import android.util.Log
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.app.AppCompatDelegate.*
import androidx.lifecycle.ViewModelProvider
import okhttp3.MultipartBody
import okhttp3.RequestBody
import online.fatbook.fatbookapp.callback.ResultCallback
import online.fatbook.fatbookapp.core.user.User
import online.fatbook.fatbookapp.databinding.ActivitySplashBinding
import online.fatbook.fatbookapp.network.LoginResponse
import online.fatbook.fatbookapp.network.service.RetrofitFactory
import online.fatbook.fatbookapp.repository.AuthenticationRepository
import online.fatbook.fatbookapp.ui.viewmodel.UserViewModel
import online.fatbook.fatbookapp.util.Constants.SP_TAG
import online.fatbook.fatbookapp.util.Constants.SP_TAG_DARK_MODE
import online.fatbook.fatbookapp.util.Constants.SP_TAG_PASSWORD
import online.fatbook.fatbookapp.util.Constants.SP_TAG_USERNAME
import org.apache.commons.lang3.StringUtils

class SplashActivity : AppCompatActivity() {
    private var _binding: ActivitySplashBinding? = null
    private val binding get() = _binding!!
    private val repository by lazy { AuthenticationRepository() }

    private var username: String? = null
    private var password: String? = null

    private lateinit var sharedPreferences: SharedPreferences

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        _binding = ActivitySplashBinding.inflate(layoutInflater)
        setContentView(binding.root)

        loadSharedPreferences()
    }

    private fun loadSharedPreferences() {
        sharedPreferences =
            getSharedPreferences(SP_TAG, MODE_PRIVATE)
        setDefaultNightMode(
            if (sharedPreferences.getBoolean(SP_TAG_DARK_MODE, false)) {
                MODE_NIGHT_YES
            } else {
                MODE_NIGHT_NO
            }
        )

        username = sharedPreferences.getString(SP_TAG_USERNAME, StringUtils.EMPTY)
        password = sharedPreferences.getString(SP_TAG_PASSWORD, StringUtils.EMPTY)
        login()
    }

    private fun login() {
        val request: RequestBody = MultipartBody.Builder().setType(MultipartBody.FORM)
            .addFormDataPart("username", username.toString())
            .addFormDataPart("password", password.toString()).build()
        repository.login(request, object : ResultCallback<LoginResponse> {
            override fun onResult(value: LoginResponse?) {
                if (value != null) {
                    RetrofitFactory.updateJWT(
                        value.access_token.toString(),
                        value.username.toString()
                    )
                    startMainScreen(value.username.toString(), password.toString())
                } else {
                    logout()
                }
            }

            override fun onFailure(value: LoginResponse?) {
                logout()
            }
        })
    }

    private fun logout() {
        sharedPreferences.edit().putString(SP_TAG_USERNAME, StringUtils.EMPTY)
        sharedPreferences.edit().putString(SP_TAG_PASSWORD, StringUtils.EMPTY)
        startMainScreen("", "")
    }

    private fun startMainScreen(username: String, password: String) {
        val intent: Intent
        if (!StringUtils.isEmpty(username) || !StringUtils.isEmpty(password)) {
            intent = Intent(this@SplashActivity, MainActivity::class.java)
            intent.putExtra(SP_TAG_USERNAME, username)
            intent.putExtra(SP_TAG_PASSWORD, password)
            if (this.intent.action != null && this.intent.action != "android.intent.action.MAIN") {
                intent.action = this.intent.action
            }

        } else {
            intent = Intent(this@SplashActivity, AuthenticationActivity::class.java)
        }
        startActivity(intent)
        finish()
    }

    override fun onDestroy() {
        super.onDestroy()
        _binding = null
    }
}